dask/layers.py

```diff
@@ -846,3 +867,145 @@ def _construct_graph(self, deserializing=False):
             dsk[(self.name, i)] = (concat_func, _concat_list)
 
         return dsk
+
+
+class MaterializedIODeps(BlockwiseIODeps):
+    """Partially-materialized IO-Dep Wrapper
+
+    This class will always call `serialize` on each element of the
+    IO dependencies during distributed execution. Therefore, the
+    IO function will need to use `deserialize` if it is passed
+    a tuple with "serialized" as the first element.
+    """
+
+    def __init__(self, parts):
+        self.parts = parts
+
+    def __getitem__(self, idx: tuple):
+        return self.parts[idx]
+
+    @classmethod
+    def __dask_distributed_pack__(cls, cls_path: str, parts: dict):
+        from distributed.protocol import serialize
+
+        packed_parts = {}
+        for k, v in parts.items():
+            packed_parts[k] = ("serialized",) + serialize(v)
+        return (cls_path, packed_parts)
```

Comment:
@ian-r-rose - This is one possibly way to deal with serialization/deserialization.  Here, in the pack stage, we can just call `serialize` on every element of  list/tuple/dict.  Then, in the io function, we can just check if the possibly-serialized argument (or tuple of arguments) is passed as a tuple with the first element being `"serialized"` - Although, I know you don't like magic strings :)

Reply:
It's true, I don't like magic strings :)

Over in #7417 I guarded a `dask.distributed` import by checking if the relevant object is arraylike (the idea being, if it's not arraylike we should deserialize), but I suspect it's possible to break that

Reply:
Yeah - I think that makes sense in the context of your PR.  

The goal here is to handle a very general case.  The elements we are serializing are typically tuples with one or more element that cannot be handled in msgpack.  Rather than isolating the problematic sub-element(s), we just call `serialize` on the entire element of `parts`.

Since we are either serializing everything or nothing from the perspective of the IO function, the deserialization logic is pretty simple.